File: lightmes-miniapp/src/api/admin/mold.ts
import { apiGet, apiPost, apiPut, apiDel } from '@/api/request'

export type MoldOut = {
  id: number
  code: string
  name: string
  model: string | null
  mold_type: string
  workshop: string | null
  status: string
  sku_id: number | null
  sku_code: string | null
  sku_name: string | null
  expected_lifespan: number | null
  current_shots: number
  purchase_date: string | null
  last_maintenance_date: string | null
  next_maintenance_date: string | null
  maintenance_interval_shots: number | null
  remark: string | null
  created_at: string
  updated_at: string
}

export type MoldMaintenanceLog = {
  id: number
  maintenance_type: string
  description: string | null
  shots_at_maintenance: number | null
  checked_by: number
  created_at: string
}

export type ProcessBinding = {
  id: number
  process_id: number
  process_name: string | null
}

export const MOLD_TYPES: Record<string, string> = {
  injection: '注塑', die_casting: '压铸', stamping: '冲压',
}

export const MOLD_STATUS_LABELS: Record<string, string> = {
  active: '正常', repair: '维修中', retired: '已退役',
}

export function lifePercent(m: MoldOut): number {
  if (!m.expected_lifespan || m.expected_lifespan <= 0) return 0
  return Math.round((m.current_shots / m.expected_lifespan) * 100)
}

export function lifeTag(pct: number): string {
  if (pct >= 95) return 'danger'
  if (pct >= 80) return 'warning'
  return 'success'
}

const BASE = '/admin/mold'

export const moldApi = {
  list(params?: { mold_type?: string; status?: string }) {
    return apiGet<{ items: MoldOut[] }>(BASE, params as Record<string, unknown>, true)
  },
  create(data: Record<string, unknown>) {
    return apiPost<{ id: number; code: string; name: string }>(BASE, data, true)
  },
  get(id: number) {
    return apiGet<MoldOut>(`${BASE}/${id}`, undefined, true)
  },
  update(id: number, data: Record<string, unknown>) {
    return apiPut<MoldOut>(`${BASE}/${id}`, data, true)
  },
  delete(id: number) {
    return apiDel<{ deleted: boolean }>(`${BASE}/${id}`, true)
  },
  listMaintenanceLogs(moldId: number, params?: { offset?: number; limit?: number }) {
    return apiGet<{ items: MoldMaintenanceLog[] }>(`${BASE}/${moldId}/maintenance-logs`, params as Record<string, unknown>, true)
  },
  createMaintenanceLog(moldId: number, data: { maintenance_type: string; description?: string }) {
    return apiPost<{ id: number; maintenance_type: string }>(`${BASE}/${moldId}/maintenance-logs`, data, true)
  },
  listProcessBindings(moldId: number) {
    return apiGet<{ items: ProcessBinding[] }>(`${BASE}/${moldId}/process-bindings`, undefined, true)
  },
  setProcessBindings(moldId: number, process_ids: number[]) {
    return apiPut<{ count: number }>(`${BASE}/${moldId}/process-bindings`, { process_ids }, true)
  },
}

File: backend/app/models/mold.py
"""模具/工装管理数据模型"""
from datetime import date, datetime

from sqlalchemy import Date, DateTime, ForeignKey, Integer, String, Text, UniqueConstraint, func
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.models.base import Base


class Mold(Base):
    """模具台账"""
    __tablename__ = "molds"

    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)
    code: Mapped[str] = mapped_column(String(32), nullable=False)
    name: Mapped[str] = mapped_column(String(128), nullable=False)
    model: Mapped[str | None] = mapped_column(String(64), nullable=True)
    mold_type: Mapped[str] = mapped_column(String(32), nullable=False, server_default="injection")  # injection/die_casting/stamping/other
    workshop: Mapped[str | None] = mapped_column(String(64), nullable=True)
    status: Mapped[str] = mapped_column(String(32), nullable=False, server_default="active")  # active/retired/maintenance

    sku_id: Mapped[int | None] = mapped_column(Integer, ForeignKey("skus.id", ondelete="SET NULL"), nullable=True)
    expected_lifespan: Mapped[int | None] = mapped_column(Integer, nullable=True)  # 预计寿命（模次）
    current_shots: Mapped[int] = mapped_column(Integer, nullable=False, server_default="0")  # 当前模次
    purchase_date: Mapped[date | None] = mapped_column(Date, nullable=True)
    last_maintenance_date: Mapped[date | None] = mapped_column(Date, nullable=True)
    next_maintenance_date: Mapped[date | None] = mapped_column(Date, nullable=True)
    maintenance_interval_shots: Mapped[int | None] = mapped_column(Integer, nullable=True)  # 保养间隔模次
    remark: Mapped[str | None] = mapped_column(Text, nullable=True)

    created_at: Mapped[datetime] = mapped_column(DateTime, nullable=False, server_default=func.now())
    updated_at: Mapped[datetime] = mapped_column(DateTime, nullable=False, server_default=func.now(), onupdate=func.now())

    sku = relationship("Sku")


class MoldMaintenanceLog(Base):
    """模具维保记录"""
    __tablename__ = "mold_maintenance_logs"

    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)
    mold_id: Mapped[int] = mapped_column(Integer, ForeignKey("molds.id", ondelete="CASCADE"), nullable=False, index=True)
    maintenance_type: Mapped[str] = mapped_column(String(32), nullable=False)  # preventive/corrective/inspection
    description: Mapped[str | None] = mapped_column(Text, nullable=True)
    shots_at_maintenance: Mapped[int | None] = mapped_column(Integer, nullable=True)
    checked_by: Mapped[int | None] = mapped_column(Integer, ForeignKey("users.id", ondelete="SET NULL"), nullable=True)

    created_at: Mapped[datetime] = mapped_column(DateTime, nullable=False, server_default=func.now())

    mold = relationship("Mold")


class MoldProcessBinding(Base):
    """模具与工序关联"""
    __tablename__ = "mold_process_bindings"

    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)
    mold_id: Mapped[int] = mapped_column(Integer, ForeignKey("molds.id", ondelete="CASCADE"), nullable=False, index=True)
    process_id: Mapped[int] = mapped_column(Integer, ForeignKey("processes.id", ondelete="CASCADE"), nullable=False, index=True)

    mold = relationship("Mold")
    process = relationship("Process")
